### DISCUSSION TITLE
How to create item with M2M relation with the api

### DISCUSSION THREAD

**Discussion by martinval** [directus/directus#2722] [Archive]
I have a courses table and a participants table linked by a junction table.
In the web interface I can add courses to participants and the other way around, but when I want to create a participant via api post request, the relation is not created.

The docs are quite sparse on explaining M2M relations and item creation....

My json request body looks like this:

`  [ { "firstname": "frank","lastname": "alistair", "courses": 3 } ]`

I've tried "courses": "3" / "courses": [{"id": 3 }] / "courses": [{"id": "3" }] ....either nothing happens or I get a server error.

Thanks in advance for any help

**Comment by rijkvanzanten** [comment]
A many to many is nothing more than a one-to-many to a junction table, with a many-to-one from there to the related collection. Creating an item in a many to one would be:

```json
{
  "first_name": "Frank",
  "last_name": "Alistar",
  "courses": [
    {
      "course_id": 3
    }
  ]
}
```

(assuming the field in your junction table is called `course_id`)

**Reply by rijkvanzanten** [reply]
You can than also take this deeper in case you want to edit the actual course from this participant collection:

```json
{
  "first_name": "Frank",
  "last_name": "Alistar",
  "courses": [
    {
      "course_id": {
        "id": 3,
        "title": "My New Title"
      }
    }
  ]
}
```

**Reply by martinval** [reply]
Hi @rijkvanzanten, thanks for your response, but I keep getting this server error:
```
{
    "error": {
        "code": null,
        "message": "Internal Server Error"
    }
}
```
I tried creating a course and adding an existing participant to it:
```
  [
  	 { "firstname": "frank","lastname": "alistair","email": "mike@mike.com", "phone": "06123456",
  	 "accepted_terms": "true",
  	 "courses": [
       	   {"course_id": 2}
       ]
  	 }
  	]
```
And I tried adding a participant and adding a course, but none work.
```
  [
  	 { "date": "2020-08-24",
       "time": "10:00:00",
       "title": "Test Course",
       "teacher": 2,
       "participants": [
       	   {"participant_id": 42}
       ]
  	 }
  	]
```
The simple M2O relationship of adding a teacher to a course works fine. 
I checked the field names and field types and all looks OK, and it does all work as expected when I add in both directions with the directus app, so I'm pretty confident the set-up is OK.

I also checked with having "Allow create" turned on but to no avail.

What would be the best way to debug this. The server error doesn't give me much to go on ?

**Reply by martinval** [reply]
Another thing, the junction table is called m2m_course_participants. Can that have some sort of incidence ?

**Comment by martinval** [comment]
while this isssue is not yet resolved, I am using an intermediate solution by making an additional api call and writing both id's directly into the junction table to create the relation. (in case someone else has a similar problem...)

**Comment by abu7elo** [comment]
Hi @martinval 

I have faced the same issue and i figured out how to consume the API with M2M relationship field.

In my case, I have collection post that have many category, and the category can be for many post.

So, the correct curl is:

```
curl --location --request POST 'http://directus.local/project/items/post?access_token=MyAccrssToken' \
--header 'Content-Type: application/json' \
--data-raw '{
            "title": "Hello API X3",
            "description": "<p>Test</p>",
            "user_id": 1,
            "tags": [
                "ali",
                "mazen"
            ],
            "map": {
                "lat": 31.955330136228707,
                "lng": 35.92868328094483
            },
            "category": [
		    {
		      "category_id": {
		        "id": 1
		      }
		    },
		    {
		      "category_id": {
		        "id": 2
		      }
		    }
		  ]
}'
```

where category, is the name of database field in post collection. category_id is the FK.

**Reply by martinval** [reply]
Thanks @abu7elo, this worked great, thank you. I am now able to do what I need with:
```"courses": [ {"course_id": {"id": 8} } ]```

**Reply by dev7ch** [reply]
I am having a quite the similar problem, yes i am able to save, but i experienced that i am not able to unset the saved values in the M2M field in directus admin.

@martinval can you unset saved values in Admin UI after setting it by ur mentioned method above?

**Reply by martinval** [reply]
@dev7ch I just tried that and for me it works. I can delete from both sides of the M2M relationship from with the directus app and values are removed from the junction table as they should.
It might be worth checking out if things work as expected in the demo aapp as a reference. 
Edit: just tried it on demo and works as expected.

**Reply by dev7ch** [reply]
@martinval thanks a lot, ok yes so its probably sth on my side

**Reply by rijkvanzanten** [reply]
If you want to _unset_ the related item (either in a o2m or a m2m, they're technically the same), you can simply nullify the many field on the related item:

(example: articles-categories)

```json
{
  "categories": [
    {
      "id": 15,
      "article": null
    }
  ]
}
```

If you want to delete the related item completely, you can pass a special `$delete` flag:

```json
{
  "categories": [
    {
      "id": 15,
      "$delete": true
    }
  ]
}
```

(This is the same for m2m. A m2m is nothing more than a o2m to a junction table, followed by a m2o from the junction table to a related item)

**Reply by shashwattpandeyy** [reply]
when using `$delete : true` it's not deleting the junction table row.
` await itemservice.updateOne(record.id, {
      directus_users: [
        {
          id: 12,
          "$delete": true
        },
      ],
    }); `

**Reply by rijkvanzanten** [reply]
@shashwattpandeyy This discussion is from a very old pre-release version. that $delete flag doesn't exist in the production release. It's controlled from the relationship configuration instead 👍🏻

**Comment by dev7ch** [comment]
Heya, just a notice/warning, if you send multiple times `courses: [{ course_id: { id: 8 } }]` to set a relation it will create multiple entries for the same relational id (which is wrong and this causes you can not unset it in admin UI)

So lets say you you post or patch this, 8 is a fixed id of you relational item

```
courses: [{ 
  course_id: { id: 8 } 
}]
```

but if you would like to delete its not the same id, e.g.

```
courses: [{
  id: 753 <--- this is not the fixed id 8, its the auto increment id related to 8 in the junction which needs to be extracted before
  $delete: true
}]
```

**Comment by MarkBurvs** [comment]
For anyone else trying this out, @abu7elo's answer is correct, however for subsequents calls to create more relations, it will over-write any existing ones so you'd have to make sure to read the existing ones in first, then append the new ones when writing them back.

A far easier way that I've been using is simply to create a new item in the junction table, and then the o2m and m2o fields on each collection on either side should update themselves automatically.

So for instance, I have two collections, "movies" and "actors". Movies has a "related_actors" m2m field, and vice-versa, Actors has a "related_actors" m2m field. Both of these share the same junction table called "actors_movies".

All I need to do is create a new item in "actors_movies" like so:
```
{
       "actors_id": 2,
       "movies_id": 3,
}
```
